index.ts

```diff
@@ -1,57 +1,34 @@
 import * as fs from 'fs'
 import * as path from 'path'
-import * as vm from 'vm'
 import getCacheKeyFunction from '@jest/create-cache-key-function'
-import { transformSync, Options } from '@swc/core'
+import type { Transformer } from '@jest/transform'
+import { transformSync, transform, Options } from '@swc/core'
 
-interface JestConfig {
-  transform: [match: string, transformerPath: string, options: Options][];
-  extensionsToTreatAsEsm?: string[]
-}
-
-interface JestTransformerOption {
-  config: JestConfig;
-  transformerConfig: Options;
-}
-
-/**
- * Loads closest package.json in the directory hierarchy
- */
-function loadClosestPackageJson(attempts = 1): Record<string, unknown> {
-  if (attempts > 5) {
-    throw new Error('Can\'t resolve main package.json file')
-  }
-  const mainPath = attempts === 1 ? './' : Array(attempts).join('../')
-  try {
-    return require(path.join(process.cwd(), mainPath, 'package.json'))
-  } catch (e) {
-    return loadClosestPackageJson(attempts + 1)
-  }
-}
-
-const packageConfig = loadClosestPackageJson()
-const isEsmProject = packageConfig.type === 'module'
-
-// Jest use the `vm` [Module API](https://nodejs.org/api/vm.html#vm_class_vm_module) for ESM.
-// see https://github.com/facebook/jest/issues/9430
-const supportsEsm = 'Module' in vm
-
-function createTransformer(swcTransformOpts?: Options) {
+function createTransformer(swcTransformOpts?: Options): Transformer {
   const computedSwcOptions = buildSwcTransformOpts(swcTransformOpts)
 
   return {
-    process(src: string, filename: string, jestOptions: any) {
-      if (supportsEsm) {
-        set(computedSwcOptions, 'module.type', isEsm(filename, jestOptions) ? 'es6' : 'commonjs')
-      }
+    process(src, filename, jestOptions) {
+      set(computedSwcOptions, 'module.type', jestOptions.supportsStaticESM ? 'es6' : 'commonjs')
 
       if (!computedSwcOptions.sourceMaps) {
         set(computedSwcOptions, 'sourceMaps', 'inline')
       }
 
       return transformSync(src, { ...computedSwcOptions, filename })
     },
+    processAsync(src, filename) {
+      // async transform is always ESM
+      set(computedSwcOptions, 'module.type', 'es6')
```

Comment:
might be better to use a separate set of `computedSwcOptions` from the sync one?